### PR TITLE
Use fermi sentence in qchem dipole and number functions

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
The `qchem.particle_number` and `qchem.fermionic_dipole` functions use a fermi sentence internally to represent a fermionic spin operator.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/qchem/dipole.py [hunk 1]
```diff
@@ -211,12 +219,12 @@ def _fermionic_dipole(*args):
             nd[1] = nd[1] + atomic_numbers[s] * mol.coordinates[i][1]
             nd[2] = nd[2] + atomic_numbers[s] * mol.coordinates[i][2]
 
-        f = []
+        d_ferm = []
         for i in range(3):
-            coeffs, ops = fermionic_observable(constants[i], integrals[i], cutoff=cutoff)
-            f.append((qml.math.concatenate((nd[i], coeffs * (-1))), [[]] + ops))
+            f = fermionic_observable(constants[i], integrals[i], cutoff=cutoff, fs=True)
```
Comment (lillian542) [inline pennylane/qchem/dipole.py:224]: Should `fs=True` be the default for `fermionic_observable` now? Or even the only output type?

**Comment by lillian542** [comment]
[sc-43900]

File: pennylane/qchem/dipole.py (hunk 1 above)
Comment (soranjh) [reply pennylane/qchem/dipole.py:224]: Yes. This [PR](https://github.com/PennyLaneAI/pennylane/pull/4556) removes the `fs` keyword.

File: pennylane/qchem/dipole.py [hunk 2]
```diff
@@ -188,9 +189,16 @@ def fermionic_dipole(mol, cutoff=1.0e-18, core=None, active=None):
...
+    + -0.4999999988651487 * a⁺(1) a(1)
+    + 0.82709948984052 * a⁺(1) a(3)
+    + 0.82709948984052 * a⁺(2) a(0)
+    + -0.4999999899792451 * a⁺(2) a(2)
+    + 0.82709948984052 * a⁺(3) a(1)
+    + -0.4999999899792451 * a⁺(3) a(3)
+    + 1.0 * I
     """
 
     def _fermionic_dipole(*args):
```
Comment (lillian542) [inline pennylane/qchem/dipole.py:204]: The "Returns" block in line 211 still describes returning a tuple of an array of floats and a list of lists of integers

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -5,6 +5,9 @@
 <h3>New features since last release</h3>
 
 <h3>Improvements 🛠</h3>
+* The qchem ``fermionic_dipole`` and ``particle_number`` functions are updated to use fermi 
+  sentence.
```
Comment (lillian542) [inline doc/releases/changelog-dev.md:9]:
```suggestion
* The qchem ``fermionic_dipole`` and ``particle_number`` functions are updated to use a ``FermiSentence``
```

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -5,6 +5,9 @@
...
+  [#4546](https://github.com/PennyLaneAI/pennylane/pull/4546)
```
Comment (lillian542) [inline doc/releases/changelog-dev.md:10]:
```suggestion
  [(#4546)](https://github.com/PennyLaneAI/pennylane/pull/4546)
```

**Review by lillian542 — APPROVED** [review]
A couple of small suggestions for documentation changes, but looks good to me!

File: pennylane/qchem/dipole.py [hunk 5]
```diff
@@ -188,9 +189,16 @@ def fermionic_dipole(mol, cutoff=1.0e-18, core=None, active=None):
     >>>                   [3.42525091, 0.62391373, 0.1688554]], requires_grad=True)
     >>> mol = qml.qchem.Molecule(symbols, geometry, alpha=alpha)
     >>> args = [alpha]
-    >>> coeffs, ops = fermionic_dipole(mol)(*args)[2]
-    >>> ops
-    [[], [0, 0], [0, 2], [1, 1], [1, 3], [2, 0], [2, 2], [3, 1], [3, 3]]
+    >>> fermionic_dipole(mol)(*args)[2]
```
Comment (Jaybsoni) [inline pennylane/qchem/dipole.py:192]: Should the [2] be there?

```suggestion
    >>> fermionic_dipole(mol)(*args)
```

File: pennylane/qchem/number.py [hunk 6]
```diff
@@ -60,11 +61,20 @@ def particle_number(orbitals):
     r = np.arange(orbitals)
     table = np.vstack([r, r, np.ones([orbitals])]).T
 
-    coeffs = np.array([])
-    ops = []
+    sentence = FermiSentence({})
 
     for i in table:
```
Comment (Jaybsoni) [inline pennylane/qchem/number.py:66]: what is i? maybe a more descriptive name

**Review by Jaybsoni — COMMENTED** [review]
Just some small clarifications and suggestions, happy to approve after

File: pennylane/qchem/dipole.py (hunk 5 above)
Comment (soranjh) [reply pennylane/qchem/dipole.py:192]: Yes, it has three components at three directions.

File: pennylane/qchem/dipole.py [hunk 7]
```diff
@@ -211,12 +218,12 @@ def _fermionic_dipole(*args):
             nd[1] = nd[1] + atomic_numbers[s] * mol.coordinates[i][1]
             nd[2] = nd[2] + atomic_numbers[s] * mol.coordinates[i][2]
 
-        f = []
+        d_ferm = []
         for i in range(3):
-            coeffs, ops = fermionic_observable(constants[i], integrals[i], cutoff=cutoff)
-            f.append((qml.math.concatenate((nd[i], coeffs * (-1))), [[]] + ops))
+            f = fermionic_observable(constants[i], integrals[i], cutoff=cutoff, fs=True)
+            d_ferm.append(FermiSentence({FermiWord({}): nd[i][0]}) - f)
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/dipole.py:224]: Yes.

File: pennylane/qchem/dipole.py [hunk 8]
```diff
@@ -283,7 +290,8 @@ def dipole_moment(mol, cutoff=1.0e-18, core=None, active=None):
     >>> mol = qml.qchem.Molecule(symbols, geometry, alpha=alpha)
     >>> args = [alpha]
     >>> dipole_moment(mol)(*args)[2].ops
-    [PauliZ(wires=[0]),
+    [Identity(wires=[0]),
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/dipole.py:293]: The output of the function is not changed. The initial doctsring was missing the Identity term.

File: tests/qchem/test_hamiltonians.py [hunk 9]
```diff
@@ -463,24 +463,26 @@ def circuit(*args):
...
                 return qml.expval(h_qubit)
 
             return circuit
 
         grad_jax = jax.grad(energy(mol), argnums=0)(*args)
 
         alpha_1 = np.array(
-            [[3.42515091, 0.62391373, 0.1688554], [3.42525091, 0.62391373, 0.1688554]],
-        )  # alpha[0][0] -= 0.0001
+            [[3.42425091, 0.62391373, 0.1688554], [3.42525091, 0.62391373, 0.1688554]],
```
[the remark this answers is not in the document]
Comment (soranjh) [reply tests/qchem/test_hamiltonians.py:476]: For consistency with the rest of the tests. The finite difference shift is changed from `0.0001` to `0.001`.

File: pennylane/qchem/number.py [hunk 10]
```diff
@@ -60,11 +61,20 @@ def particle_number(orbitals):
...
+                        (0, int(i[0])): "+",
+                        (1, int(i[1])): "-",
+                    }
+                ): i[2]
+            }
+        )
+
+    sentence.simplify()
+
+    return qubit_observable(sentence)
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/qchem/number.py]: The original code is more compact now.

File: pennylane/qchem/number.py (hunk 6 above)
Comment (soranjh) [reply pennylane/qchem/number.py]: The code is modified (to remove traces of the old version). This function basically implements the number operator N = \sum_i a_i^\dagger a_i, which is a simple Fermi sentence.

**Review by Jaybsoni — APPROVED** [review]
💯

**Pull request merged**